chainstate/tx-verifier/src/transaction_verifier/input_output_policy/mod.rs

```diff
@@ -0,0 +1,356 @@
+                            [] => Err(ConnectTransactionError::SpendStakeError(
+                                SpendStakeError::NoBlockRewardOutputs,
+                            )),
+                            [output] => match output {
+                                TxOutput::Transfer(..)
+                                | TxOutput::LockThenTransfer(..)
+                                | TxOutput::Burn(..)
+                                | TxOutput::CreateStakePool(..)
+                                | TxOutput::CreateDelegationId(..)
+                                | TxOutput::DelegateStaking(..) => {
+                                    Err(ConnectTransactionError::InvalidOutputTypeInReward)
+                                }
+                                TxOutput::ProduceBlockFromStake(..) => Ok(()),
+                            },
+                            _ => Err(ConnectTransactionError::SpendStakeError(
+                                SpendStakeError::MultipleBlockRewardOutputs,
+                            )),
+                        }
+                    }
+                },
+                // multiple inputs
+                _ => Err(ConnectTransactionError::SpendStakeError(
+                    SpendStakeError::ConsensusPoSError(ConsensusPoSError::MultipleKernels),
+                )),
+            }
+        }
+        None => {
+            // if no kernel input is present it's allowed to have multiple `LockThenTransfer` outputs,
+            // because this can only happen with PoW block reward.
+            let all_lock_then_transfer = reward
+                .outputs()
+                .ok_or(ConnectTransactionError::SpendStakeError(
+                    SpendStakeError::NoBlockRewardOutputs,
+                ))?
+                .iter()
+                .all(|output| match output {
+                    TxOutput::LockThenTransfer(..) => true,
+                    TxOutput::Transfer(..)
+                    | TxOutput::Burn(..)
+                    | TxOutput::CreateStakePool(..)
+                    | TxOutput::ProduceBlockFromStake(..)
+                    | TxOutput::CreateDelegationId(..)
+                    | TxOutput::DelegateStaking(..) => false,
+                });
+            ensure!(
+                all_lock_then_transfer,
+                ConnectTransactionError::InvalidOutputTypeInReward
+            );
+            Ok(())
+        }
+    }
+}
+
+/// Not all `TxOutput` combinations can be used in a transaction.
+pub fn check_tx_inputs_outputs_purposes(
+    tx: &Transaction,
+    utxo_view: &impl utxo::UtxosView,
+) -> Result<(), ConnectTransactionError> {
+    let inputs_utxos = tx
+        .inputs()
+        .iter()
+        .map(|input| {
+            utxo_view
+                .utxo(input.outpoint())
+                .map_err(|_| utxo::Error::ViewRead)?
+                .map(|u| u.output().clone())
+                .ok_or(ConnectTransactionError::MissingOutputOrSpent)
+        })
```

Comment:
This bit of code is repeated here and in `check_reward_inputs_outputs_purposes`. Factor out into a function?